### ISSUE TITLE
Full text Index

### ISSUE THREAD

**Issue by wojons** [rethinkdb/rethinkdb#2601]
Again another type of index that no one is planning on doing yet but would be a nice to have so less need for elasticsearch would be needed @danielmewes  said something like maybe find some sort of lucine connector or something but i know its way down the pipe line before this gets added.

**Comment by cserb** [comment]
+1 on that see #3140

**Comment by tiagocpontesp** [comment]
IMHO this should be explicitly mentioned in the site's MongoDB comparisons.

**Comment by danielmewes** [comment]
@tiagocpontesp Good point. I think when we last updated that comparison full-text indexes weren't yet available in MongoDB (I believe they were added with 3.2?). I opened https://github.com/rethinkdb/docs/issues/1162 .

**Comment by kav** [comment]
+1, I've got a simple use case and elastic search is overkill. I'll likely just manually create an index that concats all text fields. Less than ideal but it'll get the job done.

**Comment by admtech** [comment]
Yes, we need a Full text Index for rethinkdb. Elastic search ist so much overkill! Full text for rethinkdb is the missing link in the chain. Please add this to the primary feature list.

**Comment by oliverjanik** [comment]
Any news or plans?

**Comment by codermapuche** [comment]
Hello, this is a very simple but great usefull way to do a indexed full text search with great performance in large dataset (i tested inside full catalog of afip argentina ~5.000.000 of records) with great results, a search by name (like fulltext) ends in ~1s.
The create time of index is slow ~500 reads/writes per second, 4 hours of creation over 5M records in SSD storage, 8GB RAM and 4core Intel I5, this duplicate the size in disc (the index is very large).

The function fullText takes two params, the field of index and one limit of keywords, a very large limit can be used for long texts, this limit refear to maximun of keywords to index, see example query for calc the current maximun in your dataset.

For see what data is to indexed, see the exemple query of index preview.

For create the fulltext index, see the example query for indexCreate.

For perform a search fulltext after create index, see the example query for fulltext search.

For make a slow preview query of this aproach, see the preview query, note that you need to know some data of the subset included (example one name that is in the first results, see the index preview query for a hint). Whit this, you can test in a large dataset the results and the limitations before create the index, after that, the results are be the same but the speed change drastically.
Dont waste 4 hours creating the index if this not works for you. **:)**

I hope that it might be usefull for your.

As a side note, i think what a way like this can be included inside a rethinkdb core, it is fast and efficient and provide a basic fulltext tool that is usefull for the 90% of common cases.
Of course this is only my opinion, developers has a last word.

Greets.

DISCLAIMER: This is not really FULLTEXT, is a WORD-BASED text index, in the most of cases this might be ok, but keep this in mind.

```javascript
// -------------------------------------------------
// FULL TEXT TOOLS
// -------------------------------------------------

function fullTokens(source) {
  var tokens = source.downcase()
                     .split()
                     .distinct()
                     .map(function(tkn) { return { tkn: tkn } })
                     .orderBy(r.asc('tkn'));
  
  return tokens.map(function(token) {
    return token.merge({ idx: tokens.offsetsOf(token)(0) });
  });
}

// -------------------------------------------------

function fullText(source, limit) {
  var tokens = fullTokens(source);
  
  return tokens.concatMap(function(token1) {
    var range = r.range(token1('idx'), r.min([ limit, tokens.count() ])).coerceTo('array');
    
    return range.concatMap(function(idx) {      
      return range.map(function(aIdx) {                
        return tokens.filter(function(token2) { 
          return token2('idx').gt(r.max([ idx, token1('idx') ]))                  
                   .and(token2('idx').ne(idx.add(aIdx)))
                   .or(token2('idx').eq(token1('idx')));
        }) 
      })         
    })       
  })
  .distinct()
  .map(function(token) {
    return token.getField('tkn').coerceTo('string');            
  })
}

// -------------------------------------------------

function fullToken(source) {
  return fullTokens(source)('tkn').coerceTo('string');
}

// -------------------------------------------------
// FULL TEXT USE EXAMPLES
// -------------------------------------------------

// CURRENT KEYWORD LIMIT:
//r.db('argentina').table('contribuyentes')
//  .limit(10)
//  .max(function(row) { return row('denominacion').downcase().split().distinct().count(); })
//  .do(r.row('denominacion').downcase().split().distinct().count())
  
// -------------------------------------------------

// INDEX PREVIEW:
//r.db('argentina').table('contribuyentes')
//  .limit(10)
//  .map(function(doc) { return fullText(doc('denominacion'), 6); })
  
// -------------------------------------------------

// SEARCH PREVIEW:
//  r.db('argentina').table('contribuyentes')
//   .limit(25)
//  .map(function(doc) { return fullText(doc('denominacion'), 10); })
//  .filter(r.row.contains(fullToken(r.expr('nehuen'))))

// -------------------------------------------------
  
// INDEX CREATE:
//  r.db('argentina').table('contribuyentes')
//    .indexCreate('denominacion-fulltext', 
//      function(doc) { return fullText(doc('denominacion'), 6); }, 
//      { multi: true })
    
// -------------------------------------------------
  
// FULLTEXT SEARCH #1:
//  r.db('argentina').table('contribuyentes')
//    .getAll(fullToken(r.expr('nehuen prados')), { index: 'denominacion-fulltext' })

// FULLTEXT SEARCH #2:
//  r.db('argentina').table('contribuyentes')
//    .getAll(fullToken(r.expr('prados nehuen')), { index: 'denominacion-fulltext' })

// FULLTEXT SEARCH #3:
//  r.db('argentina').table('contribuyentes')
//    .getAll(fullToken(r.expr('prados')), { index: 'denominacion-fulltext' })

// FULLTEXT SEARCH #4:
//  r.db('argentina').table('contribuyentes')
//    .getAll(fullToken(r.expr('nehuen')), { index: 'denominacion-fulltext' })
    
// -------------------------------------------------
```

**Comment by codermapuche** [comment]
Another note, fullTokens function might be deterministic and can contains a regexp with match or an ngram transform, so with this aproach, you can build a custom tokenizer function that return better tokens for a particular case, allowing a custom fulltext index creation.

**Comment by kofalt** [comment]
Very interesting! Thank you for showing this. I have wondered for awhile if something like this was possible. This is a cool demonstration :)

Have you thought about connecting this with a rethink changefeed for "real-time" indexing?
Did you put an index on the search table?

**Comment by codermapuche** [comment]
@kofalt Yes!
```javascript
r.db('argentina').table('contribuyentes')
 .getAll(fullToken(r.expr('nehuen')), { index: 'denominacion-fulltext' })
  .changes({ includeInitial: true })
```
Works fine with changefeeds and insane fast! i update a 30 rows in a dataset of 5M records an voila! instantly! no performance issues!

**Comment by codermapuche** [comment]
I would love an api like this for the indices fulltext based in this aproach.
```javascript
r.db('...').table('...')
 .indexCreate('name', r.row('name').downcase().split().distinct(), { fulltext: true })
```
When the function to index is a custom tokenizer function.

**Comment by codermapuche** [comment]
Its enable some cool things, like fulltext search over multiple fields:

```
r.db('...').table('...')
 .indexCreate('fullname', r.row('firstName').add(' ').add(r.row('lastName')).downcase().split(), { fulltext: true })
 
r.db('...').table('...')
 .getAll({ firstName: 'nehuen', lastName: 'prados' }, { index: 'fullname' })
```

Love it!

**Comment by kofalt** [comment]
From your changefeed comment, if you run that query with `includeInitial` on an existing table, does it re-index the table, or is it using the `denominacion-fulltext` field to only index documents that haven't been seen yet? I am a little rusty on how that works.

I suppose as long as there's a timestamp field such as `modified` or a start-ID on the documents, you could use that to know what documents have to be re-indexed.

**Comment by codermapuche** [comment]
> From your changefeed comment, if you run that query with `includeInitial` on an existing table, does it re-index the table, or is it using the `denominacion-fulltext` field to only index documents that haven't been seen yet? I am a little rusty on how that works.
> 
> I suppose as long as there's a timestamp field such as `modified` or a start-ID on the documents, you could use that to know what documents have to be re-indexed.

includeInitial behavior is explained [in the manual](https://www.rethinkdb.com/docs/changefeeds/javascript/#including-initial-values).
this not have any special funcion here, i only use it because run the query in the browser ui and want to see the documents initials not only the changes, if you use this option or not depend of the custom case of use.

Any way, the content is only indexed one time.

Note that between command also works too with fulltext index and changefeeds, but do not figure any case of use of this.
